google/api_core/retry.py

```diff
@@ -223,22 +224,88 @@ def retry_target(target, predicate, sleep_generator, deadline, on_error=None):
+    usage of this class and the rest of the library.
+
+    Deadline - a fixed point in time by which a certain operation must
+    terminate. For example if a certain operaiton has a deadline
+    "2022-10-18T23:30:52.123Z" it must terminate (successfully or with an
+    error) till that time regradless of when it was started of if it has
+    ever been started at all.
+
+    Timeout - the maximum duration of time after which a certain operation
+    must terminate (successfully or with an error). The countdown begins right
+    after an operation was started. For example if an operation was started at
+    09:24:00 with timeout of 75 seconds, it must terminate not later than
+    09:25:15.
+
+    Unfortunately this class (and the api-core library as a whole) has not been
+    properly distinguishing the concepts of timeout and deadline and the
+    ``deadline`` parameter actually  means ``timeout``. That is why the
+    ``deadline`` has been deprecated and ``timeout`` should be used instead. If
+    ``deadline`` parameter is set, it will override ``timeout`` parameter, thus
+    ``retry.deadline`` should be treated as just a deprecated alias for
+    ``retry.timeout``.
+
+    In other words it is safe to assume that this class and the rest of this
+    library operates in terms of timeouts (not deadlines) unless explicitly
+    noted the usage of deadline semantics.
+
+    Now, when we have Timeout term properly defined, it is also important to
+    understand the three most common applications of the timeout concept in the
+    context of this library.
+
+    Usually the generic Timeout term may stand for one of the following actual
+    timeouts: RPC Timeout, Retry Timeout or Polling Timeout.
+
+    RPC Timeout - a value supplied by the client side to the server side such
+    that the server side knows the maximum amount of time it is expected to
+    spend handling that specifc RPC. For example, in case of a gRPC transport,
+    RPC Timeout is represented by setting "grpc-timeout" header in the HTTP2
+    request. The `timeout` property of this class normally never represents the
+    RPC Timeout as it is handled separately by the ``google.api_core.timeout``
+    module of this library.
+
+    Retry Timeout - this is the most common meaning of the ``timeout`` property
+    of this class, and it defines how long a certain RPC may be retried in case
+    an error is returned from the server.
+
+    Polling Timeout - it is similar to Retry Timeout, but defines how long the
+    client side is allowed to call polling rpc repeatedly to check a status of a
+    long running operaiton. Unlike in the retry case, the polling rpc is
+    expected to succed (its errors are supposed to be handled by the retry
+    logic). The decision if a new polling attemtp needs to be made is made
+    not based on the RPC status code but based on the status of the returned
+    status of an operation (i.e. it is higher level concept than the rpc error
+    codes).
```

Comment:
```suggestion
    Polling Timeout: defines how long the
    client side is allowed to call the polling RPC repeatedly to check a status of a
    long-running operation. Each polling RPC is
    expected to succeed (its errors are supposed to be handled by the retry
    logic). The decision as to whether a new polling attempt needs to be made is based
    not on the RPC status code but  on the status of the returned
    status of an operation. In other words: we will poll a long-running operation until the operation is done or the polling timeout expires. Each poll will inform us of the status of the operation. The poll consists of an RPC to the server that may itself be retried as per the poll-specific retry settings in case of errors. The operation-level retry settings do NOT apply to polling-RPC retries.
```

I tried to make this description a bit clearer. I think it's better but still needs some word-smithing.